### PR TITLE
Upgrade pybind11 API calls for 3.13t

### PR DESCRIPTION
**Pull request by albanD**
This is a modified version of https://github.com/pytorch/pytorch/pull/130341 that preserve support for older pybind version.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: torch/csrc/utils/python_symnode.cpp [hunk 1]
```diff
@@ -4,23 +4,53 @@ namespace torch {
...
+  return storage
+      .call_once_and_store_result([]() -> py::object {
+        return py::module::import("torch").attr("SymInt");
+      })
+      .get_stored();
+#else
   static py::handle symint_class =
       py::object(py::module::import("torch").attr("SymInt")).release();
   return symint_class;
+#endif
```
Comment (malfet) [inline torch/csrc/utils/python_symnode.cpp:19]: It feels like this code can benefit from good old `#define PYBIND_GET_TORCH_ATTR(name)` or a fancy template and this way avoid duplications

File: torch/csrc/jit/python/python_ivalue.h [hunk 2]
```diff
@@ -49,8 +49,22 @@ struct C10_EXPORT ConcretePyObjectHolder final : PyObjectHolder {
     // when using C++. The reason is unclear.
     try {
       pybind11::gil_scoped_acquire ag;
```
Comment (malfet) [inline torch/csrc/jit/python/python_ivalue.h:51]: Shouldn't this one be part of `#else` flag? I.e. `gil_safe_call_once_and_store` sounds like a function that would acquire gil automatically before the call, wouldn't it?

**Review by malfet — APPROVED** [review]

File: torch/csrc/jit/python/python_ivalue.h (hunk 2 above)
Comment (albanD) [reply torch/csrc/jit/python/python_ivalue.h:51]: gil_safe_call_once_and_store is a type, not a function. So we need to hold the gil for this

Comment (albanD) [reply torch/csrc/jit/python/python_ivalue.h:51]: Also to clarify, even in 3.13t the gil can be enabled and thus needs to be properly handled.

**Pull request closed without merging**